tools/syz-trace2syz/proggen/proggen.go

```diff
@@ -0,0 +1,554 @@
+	if straceType == nil {
+		log.Logf(1, "Generating union arg. StraceType is nil")
+	} else {
+		log.Logf(4, "Generating union arg: %s %s", syzType.TypeName, straceType.Name())
+	}
+	switch strType := straceType.(type) {
+	case *parser.Field:
+		return genUnionArg(syzType, strType.Val, ctx)
+	}
+	// Unions are super annoying because they often need to be handled case by case
+	// We might need to lookinto a matching algorithm to identify the union type that most closely
+	// matches our strace type
+
+	switch syzType.TypeName {
+	case "ipv4_addr":
+		return genIpv4Addr(syzType, straceType, ctx)
+	case "ipv6_addr":
+		return genIpv6Addr(syzType, straceType, ctx)
+	case "sockaddr_storage":
+		return genSockaddrStorage(syzType, straceType, ctx)
+	case "sockaddr_nl":
+		return genSockaddrNetlink(syzType, straceType, ctx)
+	case "ifr_ifru":
+		return genIfrIfru(syzType, straceType, ctx)
+	case "ifconf":
+		return genIfconf(syzType, straceType, ctx)
+	}
+	return prog.MakeUnionArg(syzType, genArgs(syzType.Fields[0], straceType, ctx))
+}
+
+func genBuffer(syzType *prog.BufferType, traceType parser.IrType, ctx *Context) prog.Arg {
+	if syzType.Dir() == prog.DirOut {
+		if !syzType.Varlen() {
+			return prog.MakeOutDataArg(syzType, syzType.Size())
+		}
+		switch a := traceType.(type) {
+		case *parser.BufferType:
+			return prog.MakeOutDataArg(syzType, uint64(len(a.Val)))
+		case *parser.Field:
+			return genBuffer(syzType, a.Val, ctx)
+		default:
+			switch syzType.Kind {
+			case prog.BufferBlobRand:
+				size := rand.Intn(256)
+				return prog.MakeOutDataArg(syzType, uint64(size))
+
+			case prog.BufferBlobRange:
+				max := rand.Intn(int(syzType.RangeEnd) - int(syzType.RangeBegin) + 1)
+				size := max + int(syzType.RangeBegin)
+				return prog.MakeOutDataArg(syzType, uint64(size))
+			default:
+				log.Fatalf("unexpected buffer type kind: %v. call %v arg %v", syzType.Kind, ctx.CurrentSyzCall, traceType)
+			}
+		}
+	}
+	var bufVal []byte
+	switch a := traceType.(type) {
+	case *parser.BufferType:
+		bufVal = []byte(a.Val)
+	case parser.Expression:
+		val := a.Eval(ctx.Target)
+		bArr := make([]byte, 8)
+		binary.LittleEndian.PutUint64(bArr, val)
+		bufVal = bArr
+	case *parser.PointerType:
+		val := a.Address
+		bArr := make([]byte, 8)
+		binary.LittleEndian.PutUint64(bArr, val)
+		bufVal = bArr
+	case *parser.GroupType:
+		return prog.DefaultArg(syzType)
+	case *parser.Field:
+		return genArgs(syzType, a.Val, ctx)
+	default:
+		log.Fatalf("Cannot parse type %#v for Buffer Type\n", traceType)
+	}
+	if !syzType.Varlen() {
+		size := syzType.Size()
+		for uint64(len(bufVal)) < size {
+			bufVal = append(bufVal, 0)
+		}
+		bufVal = bufVal[:size]
+	}
+	return prog.MakeDataArg(syzType, bufVal)
+}
+
+func genPtr(syzType *prog.PtrType, traceType parser.IrType, ctx *Context) prog.Arg {
+	switch a := traceType.(type) {
+	case *parser.PointerType:
+		if a.IsNull() {
+			return prog.DefaultArg(syzType)
```

Comment:
This does not generate NULL pointer, you need `prog.MakeSpecialPointerArg(typ, 0)`